### ISSUE TITLE
Teleport consuming all disk space with multipart files in /tmp directory

### ISSUE THREAD

**Issue by keitharogers** [gravitational/teleport#3182]
**What happened**:

Teleport keeps populating the /tmp directory with large 'multipart-' files. If I delete these, Teleport restarts on a loop every 15-20 seconds and recreates the files one-by-one. Eventually it stays running after having recreated all these files.

**What you expected to happen**:

I don't expect this to consume all available space and I expect them to stay deleted.

**How to reproduce it (as minimally and precisely as possible)**:

- Delete the 'multipart' files in /tmp
- Restart teleport
- Restarts on a loop until these are re-created
- Consumes all available space (in my case ~7GB)

**Environment**:
- Teleport version (use `teleport version`): Teleport v4.1.4 git:v4.1.4-0-gc487a75c go1.13.2
- Tsh version (use `tsh version`): Teleport v4.1.4 git:v4.1.4-0-gc487a75c go1.13.2
- OS (e.g. from /etc/os-release): Debian GNU/Linux 9 (stretch)

**Browser environment**

- Browser Version (for UI-related issues): N/A
- Install tools: N/A
- Others: N/A

**Relevant Debug Logs If Applicable**

N/A

**Comment by webvictim** [comment]
@keitharogers Could you please share your a Teleport config file (with any tokens etc redacted)?

**Comment by keitharogers** [comment]
Sure, here is my current running config:

`teleport:
  nodename: redacted.example.com
  auth_token: REDACTED
  auth_servers:
            - redacted.example.com:3025
  storage:
      type: dynamodb
      table_name: redactedt-dynamodb-table
      audit_events_uri:  ['file:///var/lib/teleport/audit/events', 'dynamodb://redacted-dynamodb', 'stdout://']
      region: eu-west-1
      audit_sessions_uri: "s3://redacted-s3-bucket/records"

proxy_service:
   enabled: yes
   https_cert_file: /etc/letsencrypt/live/redacted.example.com/fullchain.pem
   https_key_file: /etc/letsencrypt/live/redacted.example.com/privkey.pem
   public_addr: redacted.example.com
   web_listen_addr: 172.31.17.164:443

auth_service:
   enabled: true
   public_addr: redacted.example.com:3025
   tokens:
   - REDACTED:REDACTED

ssh_service:
   enabled: "yes"
   labels:
    environment: production
    role: jumpbox`

In an effort to rectify the issue, I had added the S3 / DynamoDB config. My previous running config (which has exactly the same issue as described) was:

`teleport:
  nodename: redacted.example.com
  auth_token: REDACTED
  auth_servers:
            - redacted.example.com:3025

proxy_service:
   enabled: yes
   https_cert_file: /etc/letsencrypt/live/redacted.example.com/fullchain.pem
   https_key_file: /etc/letsencrypt/live/redacted.example.com/privkey.pem
   public_addr: redacted.example.com
   web_listen_addr: 172.31.17.164:443

auth_service:
   enabled: true
   public_addr: redacted.example.com:3025
   tokens:
   - REDACTED:REDACTED

ssh_service:
   enabled: "yes"
   labels:
    environment: production
    role: jumpbox`

**Comment by benarent** [comment]
Ummm, maybe this a bug prior to sending data to DynamoDB / S3.

**Comment by keitharogers** [comment]
@benarent : This behaviour (as mentioned) was occurring both before *and* after switching to utilising DynamoDB / S3. I was trying to fix the bug by utilising store in AWS instead of locally but it made no difference.

Any ideas on what I can do to fix this?

**Comment by webvictim** [comment]
I've not seen this problem before - Teleport never even writes any files to `/tmp` in my experience. The `multipart` file makes it look like an S3 bug to me, although I hear what you're saying that it was occurring before you switched to using S3.

Any thoughts @klizhentas?

**Comment by keitharogers** [comment]
Yep, it was definitely happening before switching to S3 @webvictim .

**Comment by gabeio** [comment]
I'm actually seeing the exact same problem on the auth server:

```
teleport:
    nodename: unicorn

    advertise_ip: 10.0.1.86

    auth_servers:
        - 0.0.0.0:3025

auth_service:
    enabled: yes

    cluster_name: "default"

    authentication:
        type: github
        second_factor: off

    listen_addr: 0.0.0.0:3025

    public_addr: 10.0.1.86:3025

    tokens:
        - ""
        - ""

    session_recording: "proxy"

    proxy_checks_host_keys: yes

    client_idle_timeout: never

    disconnect_expired_cert: no

ssh_service:
    enabled: no

proxy_service:
    enabled: no
```

upon deleting all of the `multipart-*` files (53GBs worth) and starting teleport back up (assuming new connections coming in) there was 14GBs of created `multipart-*` files.

I am not seeing this on the proxy server, and thankfully not on the target boxes.

**Comment by gabeio** [comment]
something interesting I found at least in my setup, the auth server was being sent screen recordings *multiple* times, even within the same minute. for massive recordings (hours) this is extremely wasteful, with disk space, cpu cycles and network usage. I have added s3 as a storage service, and am waiting for the auth server to redirect everything there in hopes that this slows down the traffic to the auth server, on top of that I *believe* after a fair amount of investigation that the files will clean themselves up if the upload process completes but as hours of screen records can be GBs this can take some time to resolve itself.

**Comment by klizhentas** [comment]
This is helpful. Adding backoff and a better warning message is something we can add in the context of this issue.

**Comment by dkrutsko** [comment]
Hey there, so is there a solution to this problem? is it safe to delete these files?

**Comment by keitharogers** [comment]
I would also like to know if there is a solution to this, I'm still sat with no available space after 2 months and this thread seems to have died.

**Comment by webvictim** [comment]
I notice that we fixed a similar issue back with the release of Teleport 2.7.5, as tracked in https://github.com/gravitational/teleport/issues/2250 - I wonder if this could be something similar again.

The problem here is that several of us internally have tried to reproduce this issue but haven't had any success. If anyone on the thread is able to get into a situation where this issue is guaranteed to occur and provide detailed repro steps so we can make it happen too, we'll be able to get to it more quickly.

**Comment by dkrutsko** [comment]
I just ended up adding a cronjob to auto-delete any multipart file older than six hours. So far I haven't experienced any side effects but I'm not sure yet if this might cause any logs to be dropped.

```
0 * * * * root /usr/bin/find /tmp -name multipart-* -type f -mmin +360 -delete
```

**Comment by dkrutsko** [comment]
So I just looked at my Dynamo metrics and I noticed that I was going over my read capacity. I also went a bit over my write capacity as well. I'm not sure if this would be a contributing factor to the multipart file situation but I have increased the capacities in the meantime. Maybe you also ran into this issue.

**Comment by keitharogers** [comment]
This issue as originally raised by myself was present before even using DynamoDB, so it's not related to Dynamo. At least, not only related to usage of it. It is very annoying though. And FWIW, if you delete those files, they simply come back again later.

**Comment by keitharogers** [comment]
With all that being said, I have deleted the files again and restarted Teleport and the problem doesn't seem to be reoccurring. I can only imagine that it was trying to do something based on old information which has since been purged from the SQLite DB or something. I honestly don't know...

**Comment by gelato** [comment]
I can confirm that the issue still persists in teleport 4.2.2 - i've lost 90GB  worth of space because of this. Eventually teleport can just crash your machine and if teleport is the only way into infrastructure - this becomes frustrating (considering that we have teleport enterprise and it costs shitload of money).

**Comment by klizhentas** [comment]
We are refactoring session upload right now, will be released in 4.3.

**Comment by webvictim** [comment]
This was too big a change to get into 4.3, so it will be coming out with 4.4.
